lib/full-site-editing/template-loader.php

```diff
@@ -236,3 +241,34 @@ function gutenberg_template_render_without_post_block_context( $context ) {
 	return $context;
 }
 add_filter( 'render_block_context', 'gutenberg_template_render_without_post_block_context' );
+
+
+/**
+ * Sets the current WP_Query to return auto-draft posts.
+ *
+ * The auto-draft status indicates a new post, so allow the the WP_Query instance to
+ * return an auto-draft post for template resolution when editing a new post.
+ *
+ * @param WP_Query $wp_query Current WP_Query instance, passed by reference.
+ * @return void
+ */
+function gutenberg_resolve_template_for_new_post( $wp_query ) {
+	remove_filter( 'pre_get_posts', 'gutenberg_resolve_template_for_new_post' );
+
+	// Pages.
+	$page_id = isset( $wp_query->query['page_id'] ) ? $wp_query->query['page_id'] : null;
+
+	// Posts, including custom post types.
+	$p = isset( $wp_query->query['p'] ) ? $wp_query->query['p'] : null;
+
+	$post_id = $page_id ? $page_id : $p;
+	$post    = get_post( $post_id );
+
+	if (
+		$post &&
+		'auto-draft' === $post->post_status &&
+		current_user_can( 'edit_post', $post->ID )
```

Comment:
Adding this check here, since it's normally run for non-public posts, when the post_status isn't provided by the query, see: https://github.com/WordPress/wordpress-develop/blob/abb27bc8daf03cebb5e39abad986177d77377416/src/wp-includes/class-wp-query.php#L3097